### ISSUE TITLE
problem running GMSH, containing flexible object, with <flexcomp> using MuJoCo_MJX

### ISSUE THREAD

**Issue by federicopilot** [ZhengyiLuo/PerpetualHumanoidControl_MJX#2]
Good day, 

I am trying to simulate a flexible pipe_clamp with MuJoCo. 
I have created the object using openSCAD and have converted it into a GMSH file in order to use <flexcomp> properties so that i can simulate the flexibility properties. 
When attempting the simulation i let the object fall from a certain height letting it fall to the ground. However it appeared to be moving very slowly the viewing window: 
![pipeclamp_v1](https://github.com/user-attachments/assets/3006fedf-ecb1-4f01-becd-1dc86b17b2e2)
Here is my mujoco code: 
`import mujoco as mj
from mujoco import mjx
import numpy as np
import os
from mujoco.glfw import glfw
import time
import mujoco.viewer
import jax

# Define imported and exported files
xml_path = "pipeclamp_v1_mujocoviewer.xml"
gmsh_file = "pipeclamp_version1_vol.msh"

# Define simulation_window parameters
window_width = 1200
window_height = 900
window_title = "Rohrschelle_simulation"

# Define Simulation parameters
sim_time = 10000
timestep = 1e-5
Algorithmus= "CG" #mjx does not support PGS
integrator ="Euler" #MJX dpes not support IMPLICIT, IMPLICITFAST
light_diff = "1 1 1"
light_pos = "0 0 200"
light_direction = "0 0 -1"
print_camera_config = 0  # set to 1 to print camera config (useful for initializing view of the model)

# Define platform parameters
platform_name = "floor"
platform_startingangle = "0 0 0"
platform_pos = "0 0 0.3"
platform_size = "100 100 0.5"
platform_colour = "0.6 0.7 0 1"

# Define Rohreschelle/pipeclamp parameters
sim_obj_name = "Rohrschelle"
obj_starting_pos = "0 0 1"
obj_starting_angle = "0 90 0"
obj_colour = "0.9 0.55 0 1"

# Define Rohrschelle/pipeclamp material properties parameters
radius_of_flexelement = "0.005"
flexcomp_type = "gmsh"
dim = 3
obj_mass = 0.3
rigid = "false"
scale = "0.1 0.1 0.1"
edge_equality = "true"
damping_edge = 1
sol_ref_edge = "0.02 1"
sol_imp_edge = "0.95 0.95 0.001"
sol_ref_contact = "-1000 -1000"
sol_imp_contact = "0.95 0.95 0.001"
internal_cont = "false"
activelayers = 1
condim = 6
selfcollide = "none"
friction = "0.5 0.5 0.5" 
plugin_used = "mujoco.elasticity.solid"
poissonratio_Pa = 0.4
e_modulu_in_PaSI = 1e7
damping = 0.02

# MuJoCo model XML
MODEL_XML = f"""
    <mujoco>
        <option gravity="0 0 -9.8" timestep="{timestep}" solver="{Algorithmus}" integrator ="{integrator}" />
        <extension>
            <plugin plugin="{plugin_used}" />
        </extension>30
            
        <worldbody> 
            <geom name="{platform_name}" type="plane" pos="{platform_pos}" size="{platform_size}" euler="{platform_startingangle}" rgba="{platform_colour}"/>
            <light diffuse="{light_diff}" pos="{light_pos}" dir="{light_direction}"/>
            <body name="{sim_obj_name}" pos="{obj_starting_pos}">
                <flexcomp name="stl_file" type="{flexcomp_type}" file="{gmsh_file}" radius="{radius_of_flexelement}" dim="{dim}" euler="{obj_starting_angle}" rgba="{obj_colour}" mass="{obj_mass}" rigid="{rigid}" scale="{scale}">
                    <edge equality="{edge_equality}" damping="{damping_edge}" solref="{sol_ref_edge}" solimp="{sol_imp_edge}"/>
                    <contact internal="{internal_cont}" activelayers="{activelayers}" condim="{condim}" solref="{sol_ref_contact}" solimp="{sol_imp_contact}" selfcollide="{selfcollide}" friction="{friction}" />
                    <plugin plugin="{plugin_used}">
                        <config key="poisson" value="{poissonratio_Pa}"/> 
                        <config key="young" value="{e_modulu_in_PaSI}"/>
                        
                    </plugin>
                </flexcomp>
            </body>
        </worldbody>
    </mujoco>
    """

# Save the XML to a file
with open(xml_path, "w") as f:
    f.write(MODEL_XML)

# MuJoCo data structures
model = mj.MjModel.from_xml_path(xml_path)  # MuJoCo model
data = mj.MjData(model)  # MuJoCo data

# Run the simulation loop
with mj.viewer.launch_passive(model, data) as viewer:
  # Close the viewer automatically after 30 wall-seconds.
  start = time.time()
  while viewer.is_running() and time.time() - start < sim_time:
    step_start = time.time()

    mj.mj_step(model,data)

    viewer.sync() #Synchronizes the viewer with the current state, applying any changes made.

`
In order to speed up the simulation i decided to use MJX integration to make use of multithreading through the use of the GPU. 
However i was unable to simulate the object with MJX, even after many tries and received the following error: 

![Screenshot from 2024-08-07 16-56-47](https://github.com/user-attachments/assets/78b30b95-577c-425c-96d4-fe4ccaf2c288)

Here is the same code with mjx integration:
`import mujoco as mj
from mujoco import mjx
import numpy as np
import os
from mujoco.glfw import glfw
import time
import mujoco.viewer
import jax

# Define imported and exported files
xml_path = "pipeclamp_v1_mujocoviewer_mjx.xml"
gmsh_file = "pipeclamp_version1_vol.msh"

# Define simulation_window parameters
window_width = 1200
window_height = 900
window_title = "Rohrschelle_simulation"

# Define Simulation parameters
sim_time = 10000
timestep = 1e-5
Algorithmus= "CG" #mjx does not support PGS
integrator ="Euler" #MJX dpes not support IMPLICIT, IMPLICITFAST
light_diff = "1 1 1"
light_pos = "0 0 200"
light_direction = "0 0 -1"
print_camera_config = 0  # set to 1 to print camera config (useful for initializing view of the model)

# Define platform parameters
platform_name = "floor"
platform_startingangle = "0 0 0"
platform_pos = "0 0 0.3"
platform_size = "100 100 0.5"
platform_colour = "0.6 0.7 0 1"

# Define Rohreschelle/pipeclamp parameters
sim_obj_name = "Rohrschelle"
obj_starting_pos = "0 0 1"
obj_starting_angle = "0 90 0"
obj_colour = "0.9 0.55 0 1"

# Define Rohrschelle/pipeclamp material properties parameters
radius_of_flexelement = "0.005"
flexcomp_type = "gmsh"
dim = 3
obj_mass = 0.3
rigid = "false"
scale = "0.1 0.1 0.1"
edge_equality = "true"
damping_edge = 1
sol_ref_edge = "0.02 1"
sol_imp_edge = "0.95 0.95 0.001"
sol_ref_contact = "-1000 -1000"
sol_imp_contact = "0.95 0.95 0.001"
internal_cont = "false"
activelayers = 1
condim = 6
selfcollide = "none"
friction = "0.5 0.5 0.5" 
plugin_used = "mujoco.elasticity.solid"
poissonratio_Pa = 0.4
e_modulu_in_PaSI = 1e7
damping = 0.02

# MuJoCo model XML
MODEL_XML = f"""
    <mujoco>
        <option gravity="0 0 -9.8" timestep="{timestep}" solver="{Algorithmus}" integrator ="{integrator}" />
        <extension>
            <plugin plugin="{plugin_used}" />
        </extension>
        <worldbody>
            <geom name="{platform_name}" type="plane" pos="{platform_pos}" size="{platform_size}" euler="{platform_startingangle}" rgba="{platform_colour}"/>
              
            <body name="{sim_obj_name}" pos="{obj_starting_pos}">
                <flexcomp name="stl_file" type="{flexcomp_type}" file="{gmsh_file}" radius="{radius_of_flexelement}" dim="{dim}" euler="{obj_starting_angle}" rgba="{obj_colour}" mass="{obj_mass}" rigid="{rigid}" scale="{scale}">
                    <edge equality="{edge_equality}" damping="{damping_edge}" solref="{sol_ref_edge}" solimp="{sol_imp_edge}"/>
                    <contact internal="{internal_cont}" activelayers="{activelayers}" condim="{condim}" solref="{sol_ref_contact}" solimp="{sol_imp_contact}" selfcollide="{selfcollide}" friction="{friction}" />
                    <plugin plugin="{plugin_used}">
                        <config key="poisson" value="{poissonratio_Pa}"/> 
                        <config key="young" value="{e_modulu_in_PaSI}"/>
                        
                    </plugin>
                </flexcomp>
            </body>
        </worldbody>
    </mujoco>
    """

# Save the XML to a file
with open(xml_path, "w") as f:
    f.write(MODEL_XML)

# MuJoCo data structures
model = mj.MjModel.from_xml_path(xml_path)  # MuJoCo model
data = mj.MjData(model)  # MuJoCo data

# Create the mjx model and data structures
mjx_model = mjx.put_model(model)
mjx_data = mjx.put_data(model, data) #place them on the GPU device using MJX.

# Define a JAX-compatible step function
jit_step = jax.jit(mjx.step)
mj.mj_resetData(model, data)

# Run the simulation loop
with mj.viewer.launch_passive(model, data) as viewer:
  # Close the viewer automatically after 30 wall-seconds.
  start = time.time()
  while viewer.is_running() and time.time() - start < 30:
    step_start = time.time()

    # mj_step can be replaced with code that also evaluates
    # a policy and applies a control signal before stepping the physics.
    mjx_data = jit_step(mjx_model, mjx_data)
    data = mjx.get_data(model, mjx_data)

    viewer.sync() #Synchronizes the viewer with the current state, applying any changes made.`

i would greatly appreciate any help!
best regards!
